File: code/893/solution.cpp
#include <iostream>
using namespace std;
const long long e6 = 100 * 100 * 100;
int add_and_multi[e6 + 5], multi[e6 + 5];
int cost[10] = {6, 2, 5, 5, 4, 5, 6, 3, 7, 6};
int cal_origin(int a) {
    if (a < 10)
        return cost[a];
    return cal_origin(a / 10) + cost[a % 10];
}
long long solve(int a) {
    long long ret = 0;
    for (int i = 1; i <= a; i++) {
        int now = cal_origin(i);
        for (int j = 2; j * j <= i; j++) {
            if (i % j == 0) {
                now = min(now, multi[i / j] + multi[j] + 2);
            }
        }
        multi[i] = now;

        for (int j = 1; j + j <= i; j++) {
            now = min(now, add_and_multi[j] + add_and_multi[i - j] + 2);
        }
        add_and_multi[i] = now;
        ret += now;
    }
    return ret;
}
int main() {
    cout << solve(100) << endl;
    cout << solve(100 * 100) << endl;
    cout << solve(e6) << endl;
    return 0;
}

File: code/893/solution.go
package main /*{{{*/

import (
	"fmt"
)

var cost = [10]int{6, 2, 5, 5, 4, 5, 6, 3, 7, 6}

func cal_origin(a int) int {
	if a < 10 {
		return cost[a]
	}
	return cost[a%10] + cal_origin(a/10)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func solve(a int) int {
	ret := 0
	add_and_multi := [100*100*100 + 1]int{}
	multi := [100*100*100 + 1]int{}
	for i := 1; i <= a; i++ {
		now := cal_origin(i)
		for j := 2; j*j <= i; j++ {
			if i%j == 0 {
				now = min(now, multi[i/j]+multi[j]+2)
			}
		}
		multi[i] = now

		for j := 1; j+j <= i; j++ {
			now = min(now, add_and_multi[j]+add_and_multi[i-j]+2)
		}
		add_and_multi[i] = now
		ret += now
	}
	return ret
}

func main() {
	fmt.Println(solve(100))
	fmt.Println(solve(10000))
	fmt.Println(solve(1000000))
}
